module/zfs/dbuf.c

```diff
@@ -1218,17 +1235,105 @@ dbuf_read_done(zio_t *zio, const zbookmark_phys_t *zb, const blkptr_t *bp,
 		arc_buf_freeze(buf);
 		db->db_freed_in_flight = FALSE;
 		dbuf_set_data(db, buf);
-		db->db_state = DB_CACHED;
+		DBUF_STATE_CHANGE(db, =, DB_CACHED, "freed in flight");
 	} else {
 		/* success */
 		ASSERT(zio == NULL || zio->io_error == 0);
 		dbuf_set_data(db, buf);
-		db->db_state = DB_CACHED;
+		DBUF_STATE_CHANGE(db, =, DB_CACHED, "successful read");
 	}
 	cv_broadcast(&db->db_changed);
 	dbuf_rele_and_unlock(db, NULL, B_FALSE);
 }
 
+/*
+ * Shortcut for performing reads on bonus dbufs.  Returns whether any action
+ * was taken.
+ */
+static boolean_t
+dbuf_read_bonus(dmu_buf_impl_t *db, dnode_t *dn, uint32_t flags, int *err)
+{
+	int bonuslen, max_bonuslen;
+
+	*err = 0;
+	if (db->db_blkid != DMU_BONUS_BLKID)
+		return (B_FALSE);
+	*err = dbuf_read_verify_dnode_crypt(db, flags);
+	if (*err != 0)
+		return (B_FALSE);
+
+	bonuslen = MIN(dn->dn_bonuslen, dn->dn_phys->dn_bonuslen);
+	max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
+	ASSERT(MUTEX_HELD(&db->db_mtx));
+	ASSERT(DB_DNODE_HELD(db));
+	ASSERT3U(bonuslen, <=, db->db.db_size);
+	db->db.db_data = kmem_alloc(max_bonuslen, KM_SLEEP);
+	arc_space_consume(max_bonuslen, ARC_SPACE_BONUS);
+	if (bonuslen < max_bonuslen)
+		bzero(db->db.db_data, max_bonuslen);
+	if (bonuslen)
+		bcopy(DN_BONUS(dn->dn_phys), db->db.db_data, bonuslen);
+	DBUF_STATE_CHANGE(db, =, DB_CACHED, "bonus buffer filled");
+	return (B_TRUE);
+}
+
+static void
+dbuf_handle_partial_hole(dmu_buf_impl_t *db, dnode_t *dn)
+{
+	blkptr_t *bps = db->db.db_data;
+	int n_bps = dn->dn_indblkshift / sizeof (blkptr_t);
```

Comment:
The original code did use division here, but now would be a good time to update it to use `SPA_BLKPTRSHIFT`